boskos/ranch/priority.go

```diff
@@ -0,0 +1,219 @@
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	delete(rq.requestMap, requestID)
+	index := -1
+	for i, id := range rq.requestList {
+		if id == requestID {
+			index = i
+			break
+		}
+	}
+	if index >= 0 {
+		rq.requestList = append(rq.requestList[0:index], rq.requestList[index+1:len(rq.requestList)]...)
+		logrus.Infof("request id %s deleted", requestID)
+	}
+}
+
+// cleanup checks for all expired  or marked for deletion items and delete them.
+func (rq *requestQueue) cleanup(now time.Time) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	var newRequestList []string
+	newRequestMap := map[string]request{}
+	for _, requestID := range rq.requestList {
+		req := rq.requestMap[requestID]
+		// Checking expiration
+		if now.After(req.expiration) {
+			logrus.Infof("request id %s expired", req.id)
+			continue
+		}
+		// Keeping
+		newRequestList = append(newRequestList, requestID)
+		newRequestMap[requestID] = req
+	}
+	rq.requestMap = newRequestMap
+	rq.requestList = newRequestList
+}
+
+// getRank provides the rank of a given requestID following the order it was added (FIFO).
+// If requestID is an empty string, getRank assumes it is added last (lowest rank + 1).
+func (rq *requestQueue) getRank(requestID string, ttl time.Duration, now time.Time) int {
+	rq.update(requestID, now.Add(ttl))
+	rank := 1
+	rq.lock.RLock()
+	defer rq.lock.RUnlock()
+	for _, existingID := range rq.requestList {
+		req := rq.requestMap[existingID]
+		if now.After(req.expiration) {
+			logrus.Infof("request id %s expired", req.id)
+			continue
+		}
+		if requestID == existingID {
+			return rank
+		}
+		rank++
+	}
+	return rank
+}
+
+func (rq *requestQueue) isEmpty() bool {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	return len(rq.requestList) == 0
+}
+
+// RequestManager facilitates management of RequestQueues for a set of (resource type, resource state) tuple.
+type RequestManager struct {
+	lock     sync.Mutex
+	requests map[interface{}]*requestQueue
+	ttl      time.Duration
+	stopGC   context.CancelFunc
+	wg       sync.WaitGroup
+	// For testing only
+	now func() time.Time
+}
+
+// NewRequestManager creates a new RequestManager
+func NewRequestManager(ttl time.Duration) *RequestManager {
+	return &RequestManager{
+		requests: map[interface{}]*requestQueue{},
+		ttl:      ttl,
+		now:      time.Now,
+	}
+}
+
+func (rp *RequestManager) cleanup(now time.Time) {
+	rp.lock.Lock()
+	defer rp.lock.Unlock()
+	for key, rq := range rp.requests {
+		logrus.Infof("cleaning up %v request queue", key)
+		rq.cleanup(now)
+		if rq.isEmpty() {
+			delete(rp.requests, key)
+		}
+	}
+}
+
+// StartGC starts a goroutine that will call cleanup every gcInterval
+func (rp *RequestManager) StartGC(gcPeriod time.Duration) {
+	ctx, stop := context.WithCancel(context.Background())
+	rp.stopGC = stop
+	tick := time.Tick(gcPeriod)
+	go func() {
+		logrus.Info("starting cleanup go routine")
+		defer logrus.Info("exiting cleanup go routine")
+		rp.wg.Add(1)
```

Comment:
@cjwagner motto: don't `Add` and `Done` in the same goroutine, do the `Add` in the main thread

Reply:
done